cmd/gcp-controller-manager/dpwi/pods/eventhandler.go

```diff
@@ -0,0 +1,156 @@
+}
+
+// NewEventHandler creates a new pod eventHandler.
+func NewEventHandler(informer coreinformers.PodInformer, verifier *serviceaccounts.Verifier, syncer Syncer) (*eventHandler, error) {
+	eh := &eventHandler{
+		podIndexer: informer.Informer().GetIndexer(),
+		hasSynced:  informer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), "pod-event-queue"),
+		verifier: verifier,
+		syncer:   syncer,
+	}
+	informer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.onPodAdd,
+		UpdateFunc: eh.onPodUpdate,
+		DeleteFunc: eh.onPodDelete,
+	}, eventHandlerResyncPeriod)
+	return eh, nil
+}
+
+func (eh *eventHandler) onPodAdd(obj interface{}) {
+	eh.Enqueue(obj)
+}
+
+func (eh *eventHandler) onPodUpdate(obj, oldObj interface{}) {
+	eh.Enqueue(obj)
+}
+
+func (eh *eventHandler) onPodDelete(obj interface{}) {
+	eh.Enqueue(obj)
+}
+
+// Enqueue puts a pod event into the pods workerqueue.
+func (eh *eventHandler) Enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Errorf("internal error. Couldn't get key for Pod %+v: %v", obj, err)
+		return
+	}
+	eh.queue.Add(key)
+}
+
+// Run starts a number of workers to process pods events. The workers will
+// continue to run and process items from the workqueue until receiving stopCh.
+func (eh *eventHandler) Run(workers int, stopCh <-chan struct{}) {
+	for i := 0; i < workers; i++ {
+		go wait.Until(eh.work, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (eh *eventHandler) work() {
+	for eh.processNext() {
+	}
+}
+
+func (eh *eventHandler) processNext() bool {
+	key, quit := eh.queue.Get()
+	if quit {
+		return false
+	}
+	defer eh.queue.Done(key)
+
+	err := eh.process(key.(string))
+	if err != nil {
+		klog.Warningf("Requeuing %q due to %v", key, err)
+		eh.queue.AddRateLimited(key)
+		return true
+	}
+	eh.queue.Forget(key)
+	return true
+}
+
+// process() processes a pod event. If the KSA of the pod has a verified non-empty
+// GSA, it sends a node sync event.
+func (eh *eventHandler) process(key string) error {
+	o, exists, err := eh.podIndexer.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to get Pod %q: %v", key, err)
+	}
+	if !exists { // pod removal event
+		// Do nothing. Wait for another pod event or re-sync.
+		return nil
+	}
+	pod, ok := o.(*core.Pod)
+	if !ok {
+		return fmt.Errorf("invalid pod object from key %q: %#v", key, o)
+	}
+
+	ksa := serviceaccounts.ServiceAccount{
+		Namespace: pod.ObjectMeta.Namespace,
+		Name:      pod.Spec.ServiceAccountName,
+	}
+	gsa, err := eh.verifier.VerifiedGSA(ksa)
+	if err != nil {
+		return fmt.Errorf("failed to get verified GSA for ksa %v: %v", ksa, err)
+	}
+	if gsa == "" {
+		return nil
+	}
+	node := pod.Spec.NodeName
+	if node == "" {
+		return fmt.Errorf("empty node for pod %v", key)
```

Comment:
I think pods start out with an empty nodename?

We probably shouldn't treat this case as an error.  When the scheduler modifies the pod to assign it to a node, then we will reprocess the pod.